Feature/pandora modular shower reconstruction

etyley opened the pull request:
Pull request for moving TRACS from larreco to PandoraModularShowerCreation to fix dependency issues. Specifically, moving code from larreco to larpandora to allow for dependency on larpandoracontent which is required for the pandora sliding track fitting tool. I also took this opportunity to fix a couple of bugs, add some functionality and tidy the code base slightly:
 
- Store FindManyP objects in shower element holder to avoid constructing/deconstructing them many times per event which led to performance issues in cosmic events
- Make the tools inherit fcl parameters from the module, so they do not need to be set for every tool individually
- Renamed tools to be more descriptive and coherent
- Numerous small bug fixes

In order to run you also need https://github.com/etyley/larreco-1/tree/feature/etyley_PandoraShower to remove TRACS from larreco (A seperate pull request will be made), however it should build without the larreco changes. 

PandoraModularShowerCreation will exist along slide PandoraShowerCreation so should be a non breaking change. Experiments will have to opt in to use the modular shower creation, people have been identified to handle this for SBND, ICARUS and DUNE FD and feature branches for the experiment specific repos will be made available. 

There will be a presentation with further details in the LArSoft Co-ordination meeting, most likely on August 11th.

etyley committed:
First build of TRACS to LArPandoraModularShowerCreation

etyley committed:
First build of TRACS to LArPandoraModularShowerCreation

etyley committed:
Added fcls, tidied and updated dependancies to remove rouge larreco tracs reference

etyley committed:
Added fcls, tidied and updated dependancies to remove rouge larreco tracs reference

etyley committed:
Removed tests and updated fcls so it runs

etyley committed:
Removed tests and updated fcls so it runs

etyley committed:
Updated fcl parameters and naming

etyley committed:
Updated fcl parameters and naming

etyley committed:
Added fVerbose

etyley committed:
Added fVerbose

etyley committed:
Fixed fcl updates so it runs

etyley committed:
Fixed fcl updates so it runs

etyley committed:
Updated errors, warnings and exceptions to have correct names

etyley committed:
Updated errors, warnings and exceptions to have correct names

etyley committed:
Updated errors, warnings and exceptions to have correct names 2: Algs

etyley committed:
Updated errors, warnings and exceptions to have correct names 2: Algs

etyley committed:
Tidy and removed unused fcl params

etyley committed:
Tidy and removed unused fcl params

etyley committed:
Fixed bug in sliding dEdx tool MinAngleToWire cut

etyley committed:
Fixed bug in sliding dEdx tool MinAngleToWire cut

etyley committed:
Fixed shower center bug for charge weighting and changed default behaviour to better match pandora shower

etyley committed:
Fixed shower center bug for charge weighting and changed default behaviour to better match pandora shower

etyley committed:
Renamed tools and tidied fcls

etyley committed:
Renamed tools and tidied fcls

etyley committed:
Do not save output if it is all -999

etyley committed:
Do not save output if it is all -999

etyley committed:
Fixed error messages and made it build

etyley committed:
Fixed error messages and made it build

etyley committed:
Fixed capitalisation and removed unused fcl param

etyley committed:
Fixed capitalisation and removed unused fcl param

etyley committed:
Tidy up

etyley committed:
Tidy up

etyley committed:
Add Eigen as a dependancy

etyley committed:
Updated modular shower module fcl param inheratance to work for cheating alg

etyley committed:
Put debug message inside verbose check

etyley committed:
Made error values consistent across tools. Changed to make max/min use numeric limits

etyley committed:
Set default values for uninitialised variables

etyley committed:
Fixed assumption of 3 planes, use geometry service to check the dEdx and Energy vectors are the expected size

etyley committed:
Cleaned up CMakeLists

etyley committed:
Tidied includes and module descriptions

etyley committed:
Fixed CMakeLists so it builds without larreco checked out locally

FNALbuild commented:
A new Pull Request was created by  @etyley (Edward Tyley) for develop.

It involves the following packages:

larpandora


@LArSoft/level-1-managers, @LArSoft/level-2-managers can you please review it and eventually sign? Thanks.

cms-bot commands are listed <a href="https://cdcvs.fnal.gov/redmine/projects/larsoft/wiki/Pull_request_comments_that_trigger_CI_actions">here</a>

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//649/console

etyley committed:
Fixed getting name of templated type to build with clang

FNALbuild commented:
Pull request #10 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//652/console

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/CMakeLists.txt:
```diff
@@ -0,0 +1,24 @@
+  ${ART_FRAMEWORK_CORE}
+  ${ART_FRAMEWORK_PRINCIPAL}
+  ${ART_FRAMEWORK_SERVICES_REGISTRY}
+  ${ART_FRAMEWORK_SERVICES_BASIC}
```
The `${ART_FRAMEWORK_SERVICES_BASIC}` CMake variable doesn't exist anymore.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/CMakeLists.txt:
```diff
@@ -0,0 +1,15 @@
+include_directories( $ENV{PANDORA_INC} )
+include_directories( $ENV{LARPANDORACONTENT_INC} )
```
Are the include directories not available just through the UPS setup and CMake's find_ups_package calls from larpandora/CMakeLists.txt?

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerProduedPtrsHolder.hh:
```diff
@@ -0,0 +1,631 @@
+//###################################################################
+//### Name:        ShowerProduedPtrsHolder                        ###
```
While you're moving these files around, you might as well fix the spelling error in this file name.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/LArPandoraModularShowerCreation_module.cc:
```diff
@@ -0,0 +1,480 @@
+//################################################################################
+//### Name: LArPandoraModularShowerCreation                                    ###
```
No need for LArPandora to be part of the name of this module as it is already in the LArPandora package.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+#include <cxxabi.h>
+
+namespace reco {
+  namespace shower {
```
C++17 allows you to abbreviate this as:
```.cpp
namespace reco::shower {
  ...
}
```

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+    //Return the type as a string.
+    std::string GetType() override {
+      int status = -9;
+      return abi::__cxa_demangle(typeid(element).name(),NULL,NULL,&status);
```
Encourage you to use:
```.cpp
#include "cetlib_except/demangle.h"
...
return cet::demangle_symbol(typeid(elemenet).name());
```

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+        if(showerproperties.find(Name) != showerproperties.end()){
+          if(showerproperties[Name]->CheckShowerElement()){
+            reco::shower::ShowerElementAccessor<T> *showerprop = dynamic_cast<reco::shower::ShowerElementAccessor<T> *>(showerproperties[Name].get());
+            if(showerprop == NULL){
```
Recommend replacing all uses of `NULL` with `nullptr`.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+  }
+}
+
+class reco::shower::ShowerElementBase {
```
To the extent possible, all member functions should be `const`-qualified:
```.diff
- virtual int foo();
+ virtual int foo() const;
```
I suggest going through each of the member functions to determine which ones can be marked `const`.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+    virtual bool CheckTag(){
+      throw cet::exception("ShowerElementHolder") << "Trying to check an element that is not a product" << std::endl;
+    }
+    virtual void SetCheckTag(bool& check){
```
Is the `SetCheckTag` function necessary?  Generally best to initialize at construction time.  And why is the argument `bool&`?

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+    }
+
+
+  protected:
+
+    bool elementPtr;
```
A couple issues:

- Try to avoid `protected` data members to the extent possible.  They introduce couplings between base classes and derived classes that are very difficult to untangle.
- The type of `elementPtr` is `bool`.  Perhaps a better name would be `hasElementPtr`?

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+      return abi::__cxa_demangle(typeid(element).name(),NULL,NULL,&status);
+    }
+
+  protected:
```
See earlier comment about protected data members.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+      else return false;
+    }
+
+    void Clear(){
+      elementPtr    = 0;
+    }
```
Why is Clear function necessary?

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+          }
+        }
+        throw cet::exception("ShowerElementHolder") << "Trying to get Element: " << Name << ". This element does not exist in the element holder" << std::endl;
+        return 1;
```
Unreachable line of code.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+        return eventdataproducts[Name]->CheckShowerElement();
+      }
+      return false;
+    }
```
Several inefficiencies here:

- Should pass the argument in by `const&` to avoid a needless copy.
- In the event that the element exists, you're actually searching twice.  Would be much more efficient to do:
```.cpp
auto const it = eventdataproducts.find(name);
return  it == cend(eventdataproducts) ? false : it->CheckShowerElement();
```

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+      void SetEventElement(T& dataproduct, std::string Name){
+
+        if(eventdataproducts.find(Name) != eventdataproducts.end()){
+          reco::shower::EventDataProduct<T>* eventdataprod = dynamic_cast<reco::shower::EventDataProduct<T> *>(eventdataproducts[Name].get());
```
See comment below about extra searching in a map.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Algs/ShowerElementHolder.hh:
```diff
@@ -0,0 +1,671 @@
+          return;
+        }
+        else{
+          eventdataproducts[Name] = std::unique_ptr<reco::shower::EventDataProduct<T> >(new reco::shower::EventDataProduct<T>(dataproduct));
```
Much more concise to use:
```.cpp
eventdataproducts[name] = std::make_unique<EventDataProduct<T>(dataproducts);
```

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Tools/ShowerPCADirection_tool.cc:
```diff
@@ -0,0 +1,374 @@
+      throw cet::exception("ShowerPCADirection") << "Could not get the pandora pf particles. \
+        Something is not cofingured coreectly Please give the correct pandoa module label. Stopping";
+      return 1;
+    }
```
Can be stated more concisely:
```.cpp
auto const pfpHandle = Event.getValidHandle<std::vector<recob::PFParticle>>(pFParticleLabel);
```
`getValidHandle` will throw with a descriptive error message if the product is not available.  No need to re-implement that.

This applies below as well.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Tools/ShowerPCADirection_tool.cc:
```diff
@@ -0,0 +1,374 @@
+
+    if (!fmspp.isValid()){
+      throw cet::exception("ShowerPCADirection") << "Trying to get the spacepoint and failed. Something is not configured correctly. Stopping ";
+      return 1;
```
A `return` statement should never follow a `throw` statement in the same scope--it will never be reached, and it confuses readers.  This happens in various places in this PR.

knoepfel commented on larpandora/LArPandoraEventBuilding/LArPandoraShower/Tools/ShowerPCADirection_tool.cc:
```diff
@@ -0,0 +1,374 @@
+    std::vector<art::Ptr<recob::SpacePoint> > spacePoints_pfp = fmspp.at(pfparticle.key());
+
+    //We cannot progress with no spacepoints.
+    if(spacePoints_pfp.size() == 0){return 0;}
```
Recommend:
```.cpp
if (spacePoints_pfp.empty()) return 0;
```

knoepfel reviewed (commented):
@lgarren, none of my comments should hold up the PR--I will not be able to participate in any conversations due to my vacation so no reason to delay approval on my account.

@etyley, thank you for the PR.  We've already discussed how to improve the process for next time (e.g. separating code migration from code improvements).  Here I focus on improving C++ usage and adopting robust coding paradigms.  It's possible I'm making comments on code that was already part of larreco, and I just wasn't aware of it because of the migration.

There are some straightforward changes that can be made (e.g.):

- A `return` statement should never immediately follow a `throw` statement as it will never be reached.
- All `NULL` uses should be replaced with `nullptr`.
- There are various places where the following pattern is used:
    ```.cpp
    art::Handle<MyType> h;
    if (!e.getByLabel(label_, h)) throw ...;
    ```
    A better pattern is to use the following interface:
    ```.cpp
    auto const h = e.getValidHandle<MyType>(label_);
    ```
    which will throw an exception with a diagnostic message if the product is not available.
- For all classes that inherit from a base class with a `virtual` destructor, remove the destructor declaration and definition if the definition does no work.  The compiler will generate the correct one, and it's less code to maintain.

There are some design issues worth considering:

- There are many cases where `return 0;` indicates success and `return 1;` indicates failures.  Or perhaps the other way around?  I suggest that if the only meaningful return values are 0 and 1, then the function signature should be adjusted to return a Boolean `true` or `false`, indicating success or failure, respectively.
- Using `protected` data members introduces difficult-to-maintain entanglements between base classes and derived classes.  Try to avoid them.  This can mean redesigning a significant portion of the code at times, but it pays off in the end.
- To the extent possible, all member functions should be `const`-qualified.  This makes it much easier to achieve thread-safety.  If there are member functions that cannot be `const`-qualified, try to determine if those functions are required.  For example, "setters" are generally unnecessary as it is much better to fully initialize an object at its construction.

Enough for now.

etyley committed:
Kyle PR changes part 1:
Changed getByLabel to getValidHandle (Needed to change getFindManyP)
Tidied nested namespaces
Changed to cet::demangle to get type names
Changed NULL to nullptr
Removed returns directly after exeptions (the easy ones at least)
Changed empty vector check to use empty rather than size

etyley committed:
Remove destructors of inerited tools

etyley committed:
Kyles PR changes part 2:
Updated the shower element holder and produced ptrs holder to improve
search efficiency, use make_unique and remove returns after exceptions

absolution1 commented:
I leave it down to @etyley on addressing/implementing Kyle's suggestions.
I don't foresee any other reasons to postpone/delay

absolution1 commented:
approve

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged.)

FNALbuild commented:
+tests_slf7_c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9203&builds= 
 +build

FNALbuild commented:
+tests_slf7_e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds= 
 +build

FNALbuild commented:
-sbnd_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/5284&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

FNALbuild commented:
-"lariat_failed on  for e19:prof" 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4800&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

FNALbuild commented:
-icarus_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/914&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

FNALbuild commented:
+tests_slf6_e19:py2:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds= 
 +build

FNALbuild commented:
+"argoneut on  for e19:prof" 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/4958&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

FNALbuild commented:
-"lariat_failed on  for e19:py2:prof" 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4801&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

FNALbuild commented:
+"argoneut on  for e19:py2:prof" 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/4959&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

FNALbuild commented:
+dune 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/9927&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

FNALbuild commented:
-dune_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/9928&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

FNALbuild commented:
-uboone_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5385&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

FNALbuild commented:
-uboone_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5386&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9202&builds="

etyley committed:
Improved const correctness

etyley committed:
Tidied CMakeLists

etyley committed:
Fix comparisons to make it run

FNALbuild commented:
Pull request #10 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//653/console

etyley commented:
@knoepfel Thanks very much for all of the feedback Kyle. I have implemented all of the changes you suggested.

The design issues you bring up will definitely be looked at in the future. Most of these were also present in TRACS and, after discussing with the Pandora team, we decided that we should focus on getting the code moved first before any major rewrites. 

Please let me know if there is anything else that needs clarifying or changing.

I did want to bring up the possibility of applying the Pandora .clang-format to the code. John suggested applying it to the whole larpandora repo, as opposed to just the new shower reco code, but we obviously wanted to ask the LArSoft team before doing this. If we are to do this we also had a question about version controlling the .clang-format file?

ghost commented:
trigger build

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
+tests_slf7_e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds= 
 +build

FNALbuild commented:
+tests_slf7_c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9206&builds= 
 +build

FNALbuild commented:
+tests_slf6_e19:py2:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds= 
 +build

FNALbuild commented:
-icarus_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/922&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

FNALbuild commented:
-"lariat_failed on  for e19:prof" 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4802&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

FNALbuild commented:
-"lariat_failed on  for e19:py2:prof" 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4803&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

FNALbuild commented:
+"argoneut on  for e19:prof" 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/4960&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

FNALbuild commented:
-sbnd_failed 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/5287&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

FNALbuild commented:
+"argoneut on  for e19:py2:prof" 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/4961&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

FNALbuild commented:
-dune_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/9929&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

FNALbuild commented:
-dune_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/9930&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

FNALbuild commented:
-uboone_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5387&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

FNALbuild commented:
-uboone_warning 
 for details see 
 "https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5388&builds=" 
 "for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9205&builds="

ghost commented:
CI tests are normal. Two SBND unit tests fail with the following error, which is expected since TRACS is still present in larreco in this build:

`69:     ---- Configuration BEGIN
70:       Library specification "ShowerPFPVertexStartPosition" corresponds to multiple libraries:
71:       /cvmfs/larsoft-ib.opensciencegrid.org/LArSoft/LArSoft_custom_slf7_e19_prof_20200804_021631/localProducts_LArSoft_LArSoft_lar_ci_e19_prof/larpandora/v08_13_03/slf7.x86_64.e19.prof/lib/liblarpandora_LArPandoraEventBuilding_LArPandoraShower_Tools_ShowerPFPVertexStartPosition_tool.so
72:       /cvmfs/larsoft-ib.opensciencegrid.org/LArSoft/LArSoft_custom_slf7_e19_prof_20200804_021631/localProducts_LArSoft_LArSoft_lar_ci_e19_prof/larreco/v08_32_08/slf7.x86_64.e19.prof/lib/liblarreco_ShowerFinder_ShowerTools_ShowerPFPVertexStartPosition_tool.so
73:     ---- Configuration END`

I'll approve.

ghost commented:
approve

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged.)

etyley committed:
Updates for LArSoft v09 thread safe detector property services

etyley committed:
Swap TMath, and unqualified, for std math

etyley committed:
Swapped exception to error to make ut run on cosmics

FNALbuild commented:
Pull request #10 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//667/console

etyley commented:
Hi all, I have updated the code to larsoft v09_00_00, along with the [larreco](https://github.com/LArSoft/larreco/pull/16) and [sbndcode](https://github.com/SBNSoftware/sbndcode/pull/16) pull requests. ICARUS will also need to update to v09 and test this, but I don't anticipate any issues. I think the only remaining thing to discuss is applying clang-format? But let me know if there is anything else I need to do?

etyley committed:
Update to LArSoft v09_01_00

etyley commented:
So this has now been tested in SBND and ICARUS with LArSoft v09_01_00. The feature branches are:
SBND: https://github.com/SBNSoftware/sbndcode/pull/16
ICARUS: https://github.com/SBNSoftware/icaruscode/tree/feature/howard_pandorashower_v09_01_00

FNALbuild commented:
Pull request #10 was updated. @LArSoft/level-1-managers, @LArSoft/level-2-managers can you please check and sign again.

FNALbuild commented:
The code-checks are being triggered in jenkins.

FNALbuild commented:
+code-checks  
 for details see 
 https://buildmaster.fnal.gov/buildmaster/job/GithubIntegration/job/run-pr-code-checks//685/console

absolution1 commented:
Hello.
As there were so many combinatorics (multiple larsoft and experiment branches), I manually triggered a CI run with all of the relevant branches.

larsoft (all OK): https://dbweb5.fnal.gov:8443/LarCI/app/build_detail/phase_details?build_id=lar_ci/9455&platform=Linux%203.10.0-1127.el7.x86_64&phase=trigger_builds&buildtype=slf7%20e19:prof
dune (one unrelated warning): https://dbweb5.fnal.gov:8443/LarCI/app/ns:dune/build_detail/phase_details?build_id=dune_ci/10175&platform=Linux%203.10.0-1127.el7.x86_64&phase=ci_tests&buildtype=slf7%20e19:prof
sbnd (expected warnings due to new product label names): https://dbweb5.fnal.gov:8443/LarCI/app/ns:sbnd/build_detail/phase_details?build_id=sbnd_ci/5462&platform=Linux%203.10.0-1127.el7.x86_64&phase=ci_tests&buildtype=slf7%20e19:prof
icarus (fails to build for an unrelated reason): https://dbweb5.fnal.gov:8443/LarCI/app/ns:icarus/build_detail/phase_details?build_id=icarus_ci/1070&platform=Linux%203.10.0-1127.el7.x86_64&phase=build&buildtype=slf7%20e19:prof

I am happy to go ahead with L2 approval

absolution1 commented:
approve

FNALbuild commented:
The tests are being triggered in jenkins.

FNALbuild commented:
This pull request is fully signed and it will be merged to develop and built in the next LArSoft release after it passes the integration tests. This pull request will now be reviewed by the release team before it's merged.)

FNALbuild commented:
+LArSoft tests OK on slf7 for c7:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9457&builds= 
 +build

FNALbuild commented:
+LArSoft tests OK on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9456&builds= 
 +build

FNALbuild commented:
-lariat failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:lariat/view_builds/index?offset=0&builds=lariat_ci/4925&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9456&builds=

FNALbuild commented:
-icarus tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:icarus/view_builds/index?offset=0&builds=icarus_ci/1072&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9456&builds=

FNALbuild commented:
-argoneut tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:argoneut/view_builds/index?offset=0&builds=argoneut_ci/5107&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9456&builds=

FNALbuild commented:
-sbnd tests failed on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:sbnd/view_builds/index?offset=0&builds=sbnd_ci/5464&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9456&builds=

FNALbuild commented:
-dune tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:dune/view_builds/index?offset=0&builds=dune_ci/10177&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9456&builds=

FNALbuild commented:
-uboone tests warning on slf7 for e19:prof 
 for details see 
 https://lar-ci-history.fnal.gov/LarCI/app/ns:uboone/view_builds/index?offset=0&builds=uboone_ci/5522&builds= 
 for details of the parent CI build see 
 https://lar-ci-history.fnal.gov/LarCI/app/view_builds/index?offset=0&builds=lar_ci/9456&builds=

knoepfel commented:
There are some unresolved comments, but they should not hold up this PR.  Some of the design issues should be considered in the future, and we will figure out how to move forward on the code-formatting issue.

knoepfel commented:
approve
